docs/adr/0002-docker-images-for-ruby-applications.md

```diff
@@ -0,0 +1,104 @@
+project, were essentially the defacto Docker images for GOV.UK applications.
+
+Attempting to reuse these images in the GOV.UK Docker development environment
+revealed a number of problems:
+
+1. The images [embedded the applications files][embed-files] - this meant that
+   any time a file is changed the image needed to be rebuilt, which is a
+   frequent occurrence when developing.
+2. Any time a Ruby project's Gem dependencies changed the image would need to
+   be rebuilt which [required re-installing][gem-install] every single gem,
+   which could be a slow process.
+3. As Gem dependencies were stored within a projects image no Gems could be
+   shared across applications - this meant that the initialisation process
+   for multiple GOV.UK apps could be frustratingly slow with the same Gems
+   installed in each different image, it also resulted in high disk usage.
+4. The images were only exercised in publishing-e2e-tests which ran them
+   [in production mode][production], this meant they didn't necessarily have
+   sufficient dependencies for development usage.
+
+Thus we considered whether it would be appropriate to modify these images to
+solve the problems or to create different images for GOV.UK Docker.
+
+[Docker images]: https://docs.docker.com/get-started/overview/#images
+[publishing-e2e-tests]: https://github.com/alphagov/publishing-e2e-tests
+[example-dockerfile]: https://github.com/alphagov/publishing-api/blob/8b6bc39a3f6b73c2c3e73def6a1c2a6e7687d789/Dockerfile
+[embed-files]: https://github.com/alphagov/specialist-publisher/blob/33a6f3a5d7c8b5ea79d0cd5a8232ed40391db1ef/Dockerfile#L26
+[gem-install]: https://github.com/alphagov/government-frontend/blob/51d86566de33c122aea20f30bcc0fd0d36572752/Dockerfile#L16
+[production]: https://github.com/alphagov/publishing-e2e-tests/blob/3c0e50f66c2c61a7bb98eb07cacd4fcf6233fb51/docker-compose.yml#L14
+
+## Decision
+
+We decided that we should not reuse the existing images and that we would
+resolve the problems with a different image approach for GOV.UK Docker. We
+considered the existing images to represent [standard industry
+practice][docker-rails] for web application images and that this practice
+was optimised for running a production instance of an application. We felt
+that the needs of production and development were sufficiently distinct that
+it would be simpler to have separate images than try to consolidate.
+
+We instead [created a Dockerfile][base-dockerfile] that is stored in
+GOV.UK Docker and this is used by the majority of GOV.UK Docker Ruby projects.
+This base Dockerfile contains a collection of common dependencies used by a
+significant number of projects. Projects that have additional dependencies
+have their [own Dockerfiles][whitehall-dockerfile].
+
+This enabled us to resolve the aforementioned problems. The need to embed
+files is replaced by a [shared Docker mount][home-volume] that provides the
```

Comment:
Are there two points here?

- We use a shared mount to install dependencies.
- We use a shared mount for app files.

Reply:
I've rephrased this paragraph now.

Reply:
Thanks. I realise I wasn't specific above. This is actually a separate mount:

```
${GOVUK_ROOT_DIR:-~/govuk}:/govuk:delegated
```

Note that a minor problem this solves is the need to have files for _other_ projects available, in order to test against gems (+ govuk-content-schemas). I'm not sure we need to go to this level of detail here, but I thought I'd mention it in case it fits nicely into any changes we make here.

Reply:
Oh yes, of course. Will sort.

Reply:
It's updated now.